### PR TITLE
Add CI to test scikit-image against free-threaded Python 3.13

### PR DESCRIPTION
**Pull request by andfoy**
## Description

See https://github.com/scikit-image/scikit-image/issues/7464

This PR enables testing scikit-image against free-threaded Python, this is part of the ongoing effort to test projects against the changes proposed in [PEP703](https://peps.python.org/pep-0703/)

For more information see the following related PRs in other projects:
* NumPy https://github.com/numpy/numpy/pull/26512
* SciPy https://github.com/scipy/scipy/pull/20822
* PyWavelets https://github.com/PyWavelets/pywt/pull/753
* Pandas: https://github.com/pandas-dev/pandas/pull/59058
* Pillow: https://github.com/python-pillow/Pillow/pull/8200

## Checklist

- A descriptive but concise pull request title
- [Docstrings for all functions](https://github.com/numpy/numpydoc/blob/main/doc/example.py)
- [Unit tests](https://scikit-image.org/docs/dev/development/contribute.html#testing)
- A gallery example in `./doc/examples` for new features
- [Contribution guide](https://scikit-image.org/docs/dev/development/contribute.html) is followed

## Release note

For maintainers and optionally contributors, please refer to the [instructions](https://scikit-image.org/docs/stable/development/contribute.html#documenting-changes) on how to document this PR for the release notes.

```release-note
...
```

### CODE REVIEW

**Comment by stefanv** [comment]
Thanks @andfoy, that's super helpful!

**Comment by andfoy** [comment]
Tests are now passing locally in my fork: https://github.com/andfoy/scikit-image/actions/runs/9900010082/job/27349985531

File: tools/github/before_install.sh
Comment (stefanv) [inline tools/github/before_install.sh:34]:
```suggestion
# TODO: delete when scipy, numpy, cython and pywavelets free-threaded wheels are available on PyPi
```

**Comment by stefanv** [comment]
Thanks again, this looks good to me.

Now, how do we shore up our test suite to actually create parallel faults if they are present? I suspect our current suite simply won't catch those.

**Comment by andfoy** [comment]
Well, the usual first step is to identify which classes or functions may have a shared state (usually a big culprit for concurrency issues).

I did a similar exercise throughout `scipy.interpolate` https://github.com/scipy/scipy/pull/20671 and `scipy.spatial` https://github.com/scipy/scipy/pull/20655, where we found several cases of classes that had a shared state.

However, this is a very time consuming exercise that should be conducted with planning ahead, i.e., first making sure that the most popular functionality of scikit-image is tested for concurrency first.

I don't know how much budget we can spend here from the current Quansight labs project, maybe @rgommers knows more about how we can help the project to improve its parallelism support.

**Comment by rgommers** [comment]
> Well, the usual first step is to identify which classes or functions may have a shared state (usually a big culprit for concurrency issues).

Agreed. I'm not sure that there is much shared/global state - if there is, maintainers often know where it's located. 

In scikit-image I believe there's a lot of Cython and Pythran code, and no C/C++ code (right?). Pythran code is fairly constrained in how it can be written - I believe individual numerical kernels are safe. For Cython code that's written in C-like ways, it can have thread-safety issues. I suggest auditing it quickly for odd patterns. And maybe @stefanv can suggest a couple of the most heavily used and complex functions to write some new parallel tests for? That tends to be pretty quick to do, and if they all pass that gives a good amount of confidence (plus it's a test pattern that's then easy for future contributors to emulate).

At that point I would stop unless scikit-image-specific bugs show up. I'm hoping that that won't be the case. Given the amount of Cython code and there being both known and unknown issues in Cython, I think it's perhaps more likely to hit Cython (or NumPy, or CPython) issues than problems in scikit-image itself. 

I'd also suggest opening a tracking issue to capture the plans/progress, that's easier to find that a merged PR.

**Comment by stefanv** [comment]
Should we also publish a nightly wheel for packages that test against us?

**Comment by stefanv** [comment]
Thanks for guidance on testing! We do have parallel tests for some of our Cython code, and hopefully that catches the worst of it. All of our C++ code comes via Cython, but we do have some C code for phase unwrapping (`unwrap_2d_ljmu.c` and `unwrap_3d_ljmu.c`); I'll add some parallel testing to those functions.

File: skimage/util/_map_array.py
Comment (stefanv) [inline skimage/util/_map_array.py:132]: Does this function not only handle the `copy=True` case?

File: skimage/measure/_regionprops.py
Comment (stefanv) [inline skimage/measure/_regionprops.py:1411]: What is this change for?

**Review by stefanv — APPROVED** [review]
Tentatively approved, with two (minor) questions.

File: .github/workflows/tests.yml
Comment (lagru) [inline .github/workflows/tests.yml:139]: This was just discussed in a BoF at SciPy. As I understood, `PYTHON_GIL: 0` will disable free-threading even if the binaries where build with support for it. So I am wondering how this is testing then against free-threaded? Is this just meant to test against the free-threaded build but not with free-threading enabled?

File: skimage/transform/_geometric.py
Comment (lagru) [inline skimage/transform/_geometric.py:690]: I'm a bit unclear on what this change accomplishes? Could you give me more context here? :)

**Review by lagru — COMMENTED** [review]
@andfoy, thanks for getting this process started!

**Comment by stefanv** [comment]
How does Cython's `freethreading_compatible` flag interact with the `PYTHON_GIL` variable; do we need to modify our Meson build? Maybe @ngoldbaum has thoughts.

File: skimage/transform/_geometric.py
Comment (lagru) [reply skimage/transform/_geometric.py:690]: (see also Stéfan's comment I missed https://github.com/scikit-image/scikit-image/pull/7463#discussion_r1676572874)

File: .github/workflows/tests.yml
Comment (stefanv) [reply .github/workflows/tests.yml:139]: Will disable the GIL, presumably, i.e. the opposite!

File: tools/github/before_install.sh
Comment (stefanv) [reply tools/github/before_install.sh:34]: @lagru Do you mind pushing that note change to this PR?

File: .github/workflows/tests.yml
Comment (lagru) [reply .github/workflows/tests.yml:139]: Oh, should have realized. So the free-threaded build from deadsnakes needs this flag in addition? Not including this would just run with the GIL?

Comment (ngoldbaum) [reply .github/workflows/tests.yml:139]: When all of scikit-image's dependencies are updated to declare they support being imported without the GIL and scikit-image itself is updated this can be removed. The ability to do that in Cython code was merged yesterday so we weren't able to do that for many projects until literally today. It will probably take a week or two to get all the dependencies.

File: .github/workflows/tests.yml
Comment (lagru) [inline .github/workflows/tests.yml:124]:
```suggestion
      - uses: deadsnakes/action@6c8b9b82fe0b4344f4b98f2775fcc395df45e494 # v3.1.0
```
I think this is the "wrong" hash from `main` instead of the release. I presume we also have to whitelist this action in our project config. @stefanv, I assume we just trust this action implicitly because other projects already use it? It's not clear to me who's behind this...

Comment (stefanv) [reply .github/workflows/tests.yml:124]: It's by the flake8 / pre-commit author, so I think it's fine.

File: skimage/transform/_geometric.py
Comment (stefanv) [reply skimage/transform/_geometric.py:690]: I think it's to match the standard signature for `__array__`. We just need to make sure we handle the flags correctly, or error if it's an unsupported value.

Comment (andfoy) [reply skimage/transform/_geometric.py:690]: This was causing some tests to fail, which I traced back to the NumPy documentation https://numpy.org/devdocs/user/basics.interoperability.html#the-array-method, which states that the `__array__` method should have the `__array__(self, dtype=None, copy=False)` signature

File: skimage/measure/_regionprops.py
Comment (andfoy) [reply skimage/measure/_regionprops.py:1411]: After Python 3.13, the docstrings extracted via the `__doc__` attribute now have their initial indentation offsets stripped. Thus this regex now needs to be modified in order for it to work without considering the initial indentation. See: https://docs.python.org/3.13/whatsnew/3.13.html#other-language-changes

File: .github/workflows/tests.yml
Comment (ngoldbaum) [reply .github/workflows/tests.yml:139]: See https://github.com/numpy/numpy/pull/26950 for the relevant Cython build changes in NumPy.

**Comment by andfoy** [comment]
> Should we also publish a nightly wheel for packages that test against us?

Right now, nightly free-threaded wheels are being uploaded to the Scientific Python Anaconda distribution channel. If `scikit-image` has not uploaded nightly wheels yet to the channel, we would need to open an issue requesting access like it's being done for Pillow: https://github.com/scientific-python/upload-nightly-action/issues/84

File: skimage/transform/_geometric.py
Comment (rgommers) [reply skimage/transform/_geometric.py]: The tests must be failing on the regular CPython build as well, this is unrelated to free-threading support. NumPy 2.0 should be emitting a warning for this. 

> We just need to make sure we handle the flags correctly, or error if it's an unsupported value.

Indeed. `copy=True` must now be making a copy on the scikit-image side. Perhaps best done in a separate PR?

**Comment by rgommers** [comment]
> How does Cython's `freethreading_compatible` flag interact with the `PYTHON_GIL` variable; do we need to modify our Meson build?

The `freethreading-compatible` sets the default. `PYTHON_GIL=0` overrides the default (so if an extension is not yet marked as safe, it's forced to run without the GIL anyway). Using `PYTHON_GIL=0` may be useful for the next little while, until all extension modules in all test dependencies are marked as safe and those nightly wheels have propagated.

xref https://github.com/numpy/numpy/pull/26950 for the pattern to use the new `freethreading-compatible` directive.

**Comment by stefanv** [comment]
We only have one source of C code:

```
./restoration/unwrap_2d_ljmu.c
./restoration/unwrap_3d_ljmu.c
```

If, in addition, we can guarantee that Cython and Pythran codes are safe, then I think we're good.

**Comment by andfoy** [comment]
@stefanv, thanks for the references! After looking at them, they should not have any issues regarding C module registration, since they are being included as part of other Cython source.

**Comment by andfoy** [comment]
From my side, this PR is ready

**Review by stefanv — APPROVED** [review]
Let's give it a shot!

File: skimage/transform/_geometric.py
Comment (lagru) [reply skimage/transform/_geometric.py]: See https://github.com/scikit-image/scikit-image/issues/7476.

File: tools/github/before_install.sh
Comment (lagru) [inline tools/github/before_install.sh:37]: We should keep an eye on whether this actually provides pre-releases in our CI for `FREE_THREADED_BUILD=True`. We've had issues with pip updating indirect dependencies despite the upgrade strategy being "only-if-needed".

**Review by lagru — APPROVED** [review]
Looks good otherwise provided the CI passes.

**Comment by rgommers** [comment]
Two approvals and the one CI failure is unrelated. Can this PR be merged?

**Comment by rgommers** [comment]
> Should we also publish a nightly wheel for packages that test against us?

That does seem like a good idea. In other projects we've done that as a next step after adding regular CI. And in NumPy we've now merged the default and free-threaded wheel builds; 2.1.0 should publish `cp313t` wheels to PyPI.

**Comment by lagru** [comment]
Thanks for helping us get this testing in! And sorry for the small stall at the end. It's easy to forget PRs when waiting on the CI. :see_no_evil:

**Comment by andfoy** [comment]
Let me open a followup PR to upload nightly wheels

**Pull request merged**
